### PR TITLE
Fifo: Rewrite SyncGPU

### PR DESCRIPTION
**Pull request by degasus**
The new implementation has two options:
SyncGpuMaxDistance
SyncGpuMinDistance

The MaxDistance controlls how many CPU cycles the CPU is allowed to be in front
of the GPU. Too low values will slow down extremly, too high values are as
unsynchronized and half of the games will crash.
The -MinDistance (negative) set how many cycles the GPU is allowed to be in
front of the CPU. As we are used to emulate an infinitiv fast GPU, this may be
set to any high (negative) number.

### CODE REVIEW

**Comment by comex** [comment]
GitHub seems to think this needs a rebase.

**Comment by JMC47** [comment]
This absolutely destroys performance in Mario Superstar Baseball for some reason even with very lenient settings.  Not sure what to do except not use syncgpu in it.

**Comment by Linktothepast** [comment]
Seems to work ok but there are some on screen statistics that i can't get rid of for cpu and gpu with this build.

**Comment by degasus** [comment]
@Linktothepast That's the second commit. It's to check the current syncing way to optimize this values. I'll drop it before this hits master.

**Comment by Xcedf** [comment]
Confim fixes Metroid games, increased MaxDistance value from 2M to 4M pretty good so far for Metroids recovered some speed form this, not tested other games yet, great work anyway

**Comment by degasus** [comment]
@Xcedf "from 2M"? Right now, the defaults are 200k. So did you test with 400k or 4M?

**Comment by Xcedf** [comment]
Sorry, my bad, from 200k to 400k

**Comment by degasus** [comment]
@JMC47 How do you want me to continue? Shall I drop the profiler commit so that we can merge it to master? Then we could try to figure out good default values and changing them later.

**Comment by JMC47** [comment]
I think this should be merged, and then we can implement numbers for the problematic games if absolutely need be.

**Comment by degasus** [comment]
The profile commit is dropped. SyncGPU is still disabled by default, so this shouldn't change any behavior on default settings. Let's try to merge it and to search for good values per game later.

Ready to review / merge

File: Source/Core/Core/CoreParameter.h [hunk 1]
```diff
@@ -185,6 +185,9 @@ struct SCoreStartupParameter
 	bool bDCBZOFF;
 	int iBBDumpPort;
 	bool bSyncGPU;
+	int iSyncGpuMaxDistance;
```
Comment (neobrain) [inline Source/Core/Core/CoreParameter.h]: What an excellent opportunity to start documenting obscure options while introducing them! ;)

File: Source/Core/Core/HW/SystemTimers.cpp [hunk 2]
```diff
@@ -140,8 +135,13 @@ static void SICallback(u64 userdata, int cyclesLate)
 
 static void CPCallback(u64 userdata, int cyclesLate)
 {
-	CommandProcessor::Update();
-	CoreTiming::ScheduleEvent(CP_PERIOD - cyclesLate, et_CP);
+	static u64 last_call = CoreTiming::GetTicks();
```
Comment (neobrain) [inline Source/Core/Core/HW/SystemTimers.cpp]: Are static variables a wise thing to use with regards to running two emulation sessions without restarting the program?

**Comment by degasus** [comment]
fixed the static issue

**Comment by Armada651** [comment]
LGTM

**Comment by BFG1992** [comment]
Hello guys, I'm sorry if I writing this in a wrong place and at a wrong time, but anyway.

It seems that this improvement added an issue for me. I've been experiencing an immense stuttering in F-Zero GX and Metroid Prime 2 since this commit has been merged - from 4.0-6701 to current latest, which is 4.0-6953 by now. The stuttering seems random - but it gets really immense at controller calibration and track selection in F-Zero GX, and occuring in a fixed amounts of time during starting movies in MP2.

I'm running it on my notebook Lenovo V580, it's specs are:
Intel Core i3-3120M 2.5 GHz
NVIDIA GeForce GT645M
6 Gb of RAM
Win7 SP1
No CPU-consuming apps were running during my testing.

Can you please do something about it?
If I need to fill an issue at issue tracker, just say so :)

**Comment by mimimi085181** [comment]
@BFG1992 Can you disable cpu efb access(not to confuse with efb to texture/ram) and check if the stuttering disappears then? With another one of degasus' gpu sync pr, the same happened to Mario Galaxy.

If you do disable this, make sure Dolphin does just reenable it, like it does with lots of settings.

The irc channel is a good place to get in contact with the devs btw.

**Comment by BFG1992** [comment]
@mimimi085181 It seems that "Skip EFB Access from CPU" has no effect on this stuttering.
I've tested Super Mario Galaxy just now too, and it runs fine, only a minor stuttering occuring here and there. But if I check the "Synchronize GPU thread" in game's properties window, it starts to stutter, but only a bit.

Also I've discovered that stuttering also had plagued LoZ: Wind Waker. But when I uncheck "Synchronize GPU thread" in properties, it stops. Is this thing on by default now? Is there any way to disable it in defaults? I'm asking because I like to play LoZ:WW launched via Swiss loader (Swiss' widescreen hack is better, 'cause it don't stretch game's interface), and there's no way to assign properties config to a .dol file.

I never used IRC before, but I guess I'll try, thanks :)

**Comment by AdmiralCurtiss** [comment]
There's a [current bug](https://code.google.com/p/dolphin-emu/issues/detail?id=8734) that can write per-game settings to the global settings. Check your Dolphin.ini for if SyncGPU is set to True.

**Comment by BFG1992** [comment]
@AdmiralCurtiss Oh.
That explains everything >_<
I wish I could know about it earlier, so I haven't spent two hours in testing different revs in quest to find a non-laggy one :D
Thanks.

**Comment by degasus** [comment]
@BFG1992 F-Zero is a game where SyncGPU is enabled on purpose to get this stutter. This usually happens on shader compilation, so the GPU thread did already stutter before this PR. The CPU thread however was unsynchronized, so it didn't stutter but lead to crashes. So on enabling SyncGPU, there is now a bit more stuttering, but a much more stable gameplay.

**Comment by BFG1992** [comment]
@degasus I know about stuttering related to shader compilation, but this one doesn't happen when a new effect is introduced, but rather happening continuously and seems not dependant on what's going on on screen. For example, I could enter the track selection screen in F-Zero GX and stop touching anything, and stuttering will persist, no matter how long I wait. If it supposed to be like this in Sync GPU mode, than it's sad, because for a game like F-Zero, where you're racing on crazy speeds, the fluent gameplay matters most. Stuttering just breaks all the feeling of speed, and leads to losing of fun.
I don't blame anyone or anything, I really appreciate what you guys do for the emulation scene, but I just hope there will be solution soon :)

**Comment by pauldacheez** [comment]
What you've described isn't what we'd ever refer to as a "stutter" (aside from the audio), it's just a normal goddamn slowdown in Dolphin. This isn't "supposed" to behave this way, it just does because, in layman's terms, SyncGPU has to wait in more places for stuff to finish in the right order and it's not all happening fast enough to go at full speed. Faster hardware can fix this.

**Comment by BFG1992** [comment]
@pauldacheez But I know what a normal slowdown looks like. It's a constant persistent <60/50 fps lags which comes with costant sound stuttering (after merging of the new AX-HLE engine). And, yes, I am well aware that my hardware isn't very powerful, and that F-Zero GX is a pretty demanding game. But, in the menus I was able to reach fullspeed even when Dolphin was far less optimized. And I can say that this stuttering isn't just a regular slowdown. The game runs at 60 fps, but it randomly freezes for 1-3 frames while running at 60 fps in between these freezes. That's stuttering, right?
I can make a video of this, if you want.

P.S.: Sorry for my English :)

**Comment by bb010g** [comment]
You may want to open an issue about this.
